Winterwx

ericaligo-NOAA opened the pull request:
## Description

Merging winterwx branch into develop branch

This PR adds six winter weather diagnostics to the FV3 history files to be read into the UPP. A variable precip ice density currently in the RUC LSM is moved to GFS_MP_generic_post.F90 with the density then passed into the RUC, NOAH and NOAH MP LSMs. The RUC LSM will use this precip ice density. The NOAH and NOAH MP LSMs have the option of using it. For NOAH, a new namlest variable is defined if the new density is to be used. For NOAH MP, an existing namelist variable is used with option 5 in order to use the new density (option 4 is default). The accumulated/continuous winter weather fields are passed into the restart files.

Also, the namelist parameter, nsradar_reset was more appropriately renamed nsfullradar_diag.

It also fixes a bug in the frequency over which the full radar reflectivity (water coated ice) is computed. With the bug, the full radar reflectivity was computed one time step before the top of the hour (when parameter was set to 3600s). The fix allows the full radar reflectivity to be computed at the top of the hour.

Upstream Issue(s) addressed
 - ufs-community/ufs-weather-model/issues/1449
 - ufs-community/ufs-weather-model/issues/1499

https://github.com/NOAA-EMC/UPP/issues/568

FV3 Bug fixes:
 - Closes: #437 

PRs:
 - Dependency of: ufs-community/ufs-weather-model/pull/1529
 - Depends on: ufs-community/ccpp-physics/pull/30

Results will change for any SDF that includes the RUC LSM.Results will change if lradar=.true. because of the bug fix. Results will change since UPP expects all six diagnostics to be read in and no diag_tables were updated.

Testing
No RTs were run. Code changes were tested in a slightly older verison of ufs-weather-model in RRFS as well as global configurations. New RTs will be run on hera.

[ x] hera.intel
Dependencies
None

### Issue(s) addressed

https://github.com/ufs-community/ufs-weather-model/issues/1449

ericaligo-NOAA committed:
Update .gitmodules

ericaligo-NOAA committed:
Six new winter wx diags added to output, accumulated fields added to restart file and changed nsradar_reset, radar_reset to something that makes more sense, that is, nsfullradar_diag and fullradar_diag. Also, added  new namelist parameter to use new precip ice density with the  NOAH LSM. Code changes to upp also made to include the new winter wx diags.

ericaligo-NOAA committed:
Fix typo

ericaligo-NOAA committed:
Rename winter wx diags

ericaligo-NOAA committed:
Rename winter wx diags

ericaligo-NOAA committed:
Rename winter wx diags

ericaligo-NOAA committed:
removed log file

ericaligo-NOAA committed:
remove six winter wx diags from f00 xml file

ericaligo-NOAA committed:
Merge updates

ericaligo-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into winterwx

ericaligo-NOAA committed:
Merge atmoscubed and framework

ericaligo-NOAA committed:
point to the updated UPP

ChunxiZhang-NOAA commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -1729,6 +1730,13 @@ module GFS_typedefs
     real (kind=kind_phys), pointer :: toticeb(:)     => null()   !< accumulated ice precipitation in bucket (kg/m2)
     real (kind=kind_phys), pointer :: totsnwb(:)     => null()   !< accumulated snow precipitation in bucket (kg/m2)
     real (kind=kind_phys), pointer :: totgrpb(:)     => null()   !< accumulated graupel precipitation in bucket (kg/m2)
+    real (kind=kind_phys), pointer :: frzr(:)     => null()   !< accumulated surface freezing rain (m)
```
Please follow the format of => null() in the previous row

ChunxiZhang-NOAA commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -7211,6 +7231,9 @@ subroutine diag_phys_zero (Diag, Model, linit, iauwindow_center)
     Diag%toticeb    = zero
     Diag%totsnwb    = zero
     Diag%totgrpb    = zero
+    Diag%frzrb   = zero
```
Please follow the format of = zero in the previous row.

ChunxiZhang-NOAA commented on ccpp/driver/GFS_diagnostics.F90:
```diff
@@ -1642,6 +1642,89 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
+    ExtDiag(idx)%unit = 'kg/m**2'
+    ExtDiag(idx)%mod_name = 'gfs_phys'
+    ExtDiag(idx)%cnvfac = cn_th
+    allocate (ExtDiag(idx)%data(nblks))
```
Maybe add  ExtDiag(idx)%intpl_method = 'bilinear'?

ChunxiZhang-NOAA commented:
@ericaligo-NOAA  Please update the hash for ccpp/framework.

ericaligo-NOAA commented:
Thanks, Chunxi. For these diagnostics we would want nearest neighbor. I 
was told not specifying intpl_method means the nearest neighbor will be 
used.
On 1/5/2023 2:29 PM, ChunxiZhang-NOAA wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In ccpp/driver/GFS_diagnostics.F90 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#discussion_r1062822172>:
>
> > @@ -1642,6 +1642,89 @@ subroutine GFS_externaldiag_populate (ExtDiag, Model, Statein, Stateout, Sfcprop
>         ExtDiag(idx)%data(nb)%var2 => IntDiag(nb)%pratemax(:)
>       enddo
>   
> +    idx = idx + 1
> +    ExtDiag(idx)%axes = 2
> +    ExtDiag(idx)%name = 'frzr'
> +    ExtDiag(idx)%desc = 'accumulated surface freezing rain'
> +    ExtDiag(idx)%unit = 'kg/m**2'
> +    ExtDiag(idx)%mod_name = 'gfs_phys'
> +    ExtDiag(idx)%cnvfac = cn_th
> +    allocate (ExtDiag(idx)%data(nblks))
>
> Maybe add ExtDiag(idx)%intpl_method = 'bilinear'?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#pullrequestreview-1237930902>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ALQ75MK5CUH7G2AK2U5ATW3WQ4OJ7ANCNFSM6AAAAAATMC6YIY>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

ChunxiZhang-NOAA commented:
@ericaligo-NOAA I see, thanks.

ericaligo-NOAA committed:
minor touchups

ericaligo-NOAA commented:
Here's what I have in my .gitmodules:

[submodule "ccpp/framework"]
   path = ccpp/framework
   url = https://github.com/NCAR/ccpp-framework
   branch = main

What should I change this to?

Eric
On 1/5/2023 2:32 PM, ChunxiZhang-NOAA wrote:
>
> @ericaligo-NOAA <https://github.com/ericaligo-NOAA> Please update the 
> hash for ccpp/framework.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#issuecomment-1372643662>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ALQ75MIZPPWSKVA3QL53COLWQ4OTXANCNFSM6AAAAAATMC6YIY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

ChunxiZhang-NOAA commented:
@ericaligo-NOAA Keep .gitmodules as is.  do the following steps: cd ccpp/framework;  git checkout d4a0003; cd ../..; git status; git add ccpp/framework; git commit; git push origin your branch.

ericaligo-NOAA committed:
Minor changes to typedefs

ericaligo-NOAA committed:
switch to new hash

ericaligo-NOAA commented:
Done. Thanks for the help!
On 1/5/2023 4:01 PM, ChunxiZhang-NOAA wrote:
>
> @ericaligo-NOAA <https://github.com/ericaligo-NOAA> Keep .gitmodules 
> as is. do the following steps: cd ccpp/framework; git checkout 
> d4a0003; cd ../..; git status; git add ccpp/framework; git commit; git 
> push origin your branch.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#issuecomment-1372746827>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ALQ75MORDLTUWDZH4JTKJJLWQ4ZD3ANCNFSM6AAAAAATMC6YIY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

ChunxiZhang-NOAA commented:
@ericaligo-NOAA You are very welcome!

ericaligo-NOAA committed:
Removed commented out lines related to the old computation of the precip ice density

ericaligo-NOAA commented:
Okay, I removed the commented out lines in the RUC LSM and pushed those 
changes to my branch. Hopefully I've addressed all of your 
comments/suggestions.
On 1/5/2023 4:15 PM, ChunxiZhang-NOAA wrote:
>
> @ericaligo-NOAA <https://github.com/ericaligo-NOAA> You are very welcome!
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#issuecomment-1372778264>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ALQ75MKYKCMCYCFNM5LKXJ3WQ42X7ANCNFSM6AAAAAATMC6YIY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

ChunxiZhang-NOAA commented:
@ericaligo-NOAA  Please make all your changes, then commit your changes in one commit.

ChunxiZhang-NOAA commented:
> @ericaligo-NOAA Please make all your changes, then commit your changes in one commit.

I mean next time when you make changes.

ChunxiZhang-NOAA commented:
> Okay, I removed the commented out lines in the RUC LSM and pushed those changes to my branch. Hopefully I've addressed all of your comments/suggestions.
> […](#)
> On 1/5/2023 4:15 PM, ChunxiZhang-NOAA wrote: @ericaligo-NOAA <https://github.com/ericaligo-NOAA> You are very welcome! — Reply to this email directly, view it on GitHub <[#614 (comment)](https://github.com/NOAA-EMC/fv3atm/pull/614#issuecomment-1372778264)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ALQ75MKYKCMCYCFNM5LKXJ3WQ42X7ANCNFSM6AAAAAATMC6YIY>. You are receiving this because you were mentioned.Message ID: ***@***.***>

Thank you!

ericaligo-NOAA commented:
Hi, I did a fresh clone of my branch and I'm using the latest upp and framework hashes. My compile failed, and it doesn't seem related to any of my changes. I did a fresh clone of develop and it does compile. I don't know why my branch fails with the latest framework hash. The error message is below:

Scanning dependencies of target fv3atm
[ 94%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/cpl/module_block_data.F90.o
[ 94%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/cpl/module_cplfields.F90.o
[ 94%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/FV3GFS_io.F90.o
[ 94%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/module_fv3_config.F90.o
[ 95%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/stochastic_physics/stochastic_physics_wrapper.F90.o
[ 95%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/cpl/module_cap_cpl.F90.o
[ 95%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/module_fv3_io_def.F90.o
[ 96%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/module_write_internal_state.F90.o
[ 96%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/post_nems_routines.F90.o
[ 96%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/module_write_netcdf.F90.o
[ 96%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/post_fv3.F90.o
[ 96%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/io/module_wrt_grid_comp.F90.o
[ 97%] Building Fortran object FV3/CMakeFiles/fv3atm.dir/atmos_model.F90.o
/scratch1/NCEPDEV/stmp2/Eric.Aligo/ufs-weather-model/FV3/atmos_model.F90(570): error #6633: The type of the actual argument differs from the type of the dummy argument.   [NGRIDS]
                           Atmos%ngrids, Atmos%mygrid, Atmos%pelist)
---------------------------------^
/scratch1/NCEPDEV/stmp2/Eric.Aligo/ufs-weather-model/FV3/atmos_model.F90(570): error #6633: The type of the actual argument differs from the type of the dummy argument.   [MYGRID]
                           Atmos%ngrids, Atmos%mygrid, Atmos%pelist)
-----------------------------------------------^
/scratch1/NCEPDEV/stmp2/Eric.Aligo/ufs-weather-model/FV3/atmos_model.F90(568): error #6631: A non-optional actual argument must be present when invoking a procedure with an explicit interface.   [MYGRID_ATMOS]
   call atmosphere_domain (Atmos%domain, Atmos%domain_for_read, Atmos%layout, &
--------^
/scratch1/NCEPDEV/stmp2/Eric.Aligo/ufs-weather-model/FV3/atmos_model.F90(568): error #6631: A non-optional actual argument must be present when invoking a procedure with an explicit interface.   [PELIST]
   call atmosphere_domain (Atmos%domain, Atmos%domain_for_read, Atmos%layout, &
--------^
/scratch1/NCEPDEV/stmp2/Eric.Aligo/ufs-weather-model/FV3/atmos_model.F90(570): error #6634: The shape matching rules of actual arguments and dummy arguments have been violated.   [PELIST]
                           Atmos%ngrids, Atmos%mygrid, Atmos%pelist)
-------------------------------------------------------------^
compilation aborted for /scratch1/NCEPDEV/stmp2/Eric.Aligo/ufs-weather-model/FV3/atmos_model.F90 (code 1)
make[2]: *** [FV3/CMakeFiles/fv3atm.dir/atmos_model.F90.o] Error 1
make[1]: *** [FV3/CMakeFiles/fv3atm.dir/all] Error 2
make: *** [all] Error 2
@junwang-noaa @DusanJovic-NOAA

ericaligo-NOAA committed:
Merge upstream develop branch into winterwx
Merge remote-tracking branch 'upstream/develop' into winterwx

ericaligo-NOAA committed:
commit correct hash to winterwx branch

grantfirl commented:
This isn't mentioned explicitly in the description (perhaps it is implied), but this PR is updating the commit hash of UPP to the latest develop hash, a jump of 5 months or so of commits.

WenMeng-NOAA commented:
The revision (e88dbea) of  UPP submodule is an appropriate UPP version for this PR.

grantfirl commented:
> The revision (e88dbea) of UPP submodule is an appropriate UPP version for this PR.

Thanks for the confirmation. I think that my question was in reference to the fact that updating the UPP hash presumably has a lot more changes than just what was required for @ericaligo-NOAA 's winter weather diagnostics changes. I wanted to confirm with the UFS code managers that all of the code changes in UPP represented by the update to the hash are desired/expected.

WenMeng-NOAA commented:
> This isn't mentioned explicitly in the description (perhaps it is implied), but this PR is updating the commit hash of UPP to the latest develop hash, a jump of 5 months or so of commits.

Eric recently made changes for post processing new winter weather variables in UPP repository.  To generate these variables in GRIB2, updating revision of upp submodule is needed.

grantfirl commented:
For example, the previous UPP commit hash for the fv3atm develop branch was back in August 2022: [upp @ e227247](https://github.com/NOAA-EMC/UPP/tree/e22724738fd104327fee7c3c7ffc805ccabd619f)

WenMeng-NOAA commented:
> > The revision (e88dbea) of UPP submodule is an appropriate UPP version for this PR.
> 
> Thanks for the confirmation. I think that my question was in reference to the fact that updating the UPP hash presumably has a lot more changes than just what was required for @ericaligo-NOAA 's winter weather diagnostics changes. I wanted to confirm with the UFS code managers that all of the code changes in UPP represented by the update to the hash are desired/expected.

I could help to verify the changes of UPP outputs in grib2 besides new winter weather changes.

ericaligo-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into winterwx

ericaligo-NOAA committed:
update submodule ccpp/physics

ericaligo-NOAA committed:
winter weather diagnostics related changes

grantfirl commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -1729,6 +1730,14 @@ module GFS_typedefs
+    real (kind=kind_phys), pointer :: tsnowp (:)     => null()   !< accumulated surface snowfall (m)
+    real (kind=kind_phys), pointer :: tsnowpb(:)     => null()   !< accumulated surface snowfall in bucket (m)
+    real (kind=kind_phys), pointer :: rhonewsn1(:)   => null()   !< precipitation ice density outside RUC LSM (kg/m3)
+    real (kind=kind_phys), pointer :: rhosnf(:)      => null()   !< precipitation ice density inside RUC LSM (kg/m3)
```
Since you're adding rhosnf (lsm_internal_surface_frozen_precipitation_density) to the Diag DDT, you can remove the rhofr variable in the Sfcprop DDT (and any allocations/initializations/metadata) because it is replaced by rhosnf in RUC LSM.

ericaligo-NOAA replied:
Grant,
Did you add and then remove rhosnf from GFS_typdefs.F90 at one point? I don't see it in my older versions of my branch (locally), but I see it in my latest branch suggesting I had added it based on your suggestion. No problem, I just want to make sure it should be removed. 

Also, what do you think about rhosnf in module_sf_ruclsm.F90? It's initialized to -1.e3  in the beginning of the code, redefined as rhosnfall. Rhosnfall is then passed into the SFCTMP routine. It's declared as INOUT. As you can see below, when in the if else construct, it will not be redefined if exticeden is true. When we get out of the SFCTMP routine, RHOSNFALL becomes RHOSNF, but it looks like RHOSNF doesn't do anything. Just need to be careful here. What do you think?

1078 !--- RHOSNF - density of snowfall
1079        RHOSNF(I,J)=RHOSNFALL

1162    SUBROUTINE SFCTMP (debug_print, delt,ktau,conflx,i,j,         & !--- input variables
1163                 nzs,nddzs,nroot,meltfactor,                      &
1164                 ILAND,ISOIL,IVGTYP,ISLTYP,PRCPMS,                &
1165                 NEWSNMS,SNWE,SNHEI,SNOWFRAC,                     &
1166                 exticeden,RHOSN,RHONEWSN_ex,RHONEWSN,RHOSNFALL,  &

1278              INTENT(INOUT)   ::                             DEW, &
1279                                                           EDIR1, &
1280                                                             EC1, &
1281                                                            ETT1, &
1282                                                            EETA, &
1283                                                           EVAPL, &
1284                                                         INFILTR, &
1285                                                           RHOSN, &
1286                                                        RHONEWSN, &
1287                                                       rhosnfall, &

1498       if (exticeden) then
1499         rhonewsn = rhonewsn_ex
1500       else
1501         rhonewsn=min(125.,1000.0/max(8.,(17.*tanh((276.65-Tabs)*0.15))))
1502         rhonewgr=min(500.,rhowater/max(2.,(3.5*tanh((274.15-Tabs)*0.3333))))
1503         rhonewice=rhonewsn
1504 !--- compute density of "snowfall" from weighted contribution
1505 !                 of snow, graupel and ice fractions
1506
1507         rhosnfall = min(500.,max(58.8,(rhonewsn*snowrat +  &
1508 !13mar18         rhosnfall = min(500.,max(76.9,(rhonewsn*snowrat +  &
1509                      rhonewgr*grauprat + rhonewice*icerat + rhonewgr*curat)))
1510
1511 ! from now on rhonewsn is the density of falling frozen precipitation
1512         rhonewsn=rhosnfall
1513       end if

grantfirl replied:
@ericaligo-NOAA In my branch, I repurposed the Sfcprop%rhofr variable to have the standard name of `lsm_internal_surface_frozen_precipitation_density`. In this PR branch, the Diag%rnosnf variable is used as `lsm_internal_surface_frozen_precipitation_density` instead. This is fine, but Sfcprop%rhofr is abandoned with this implementation, and should be removed.

Originally, yes, I removed Sfcprop%rhofr, but added it back in and gave it a new standard name in my branch after the discussion with Tanya.

grantfirl replied:
As long as `lsm_internal_surface_frozen_precipitation_density` exists in the metadata and the memory is allocated in fv3atm somewhere, it is fine.

ericaligo-NOAA replied:
I still believe there could be an issue below because rhosnfall is populated if exticeden is false, but it's not populated if it's true. Rhosnfall is not being used currently in the LSM, but if someone decides to use it while exticeden is true, the initial negative value would be used.
      if (exticeden) then
        rhonewsn = rhonewsn_ex
        **rhosnfall = rhonewsn** <------------------------I think this should be added for consistency.
      else
        rhonewsn=min(125.,1000.0/max(8.,(17.*tanh((276.65-Tabs)*0.15))))
        rhonewgr=min(500.,rhowater/max(2.,(3.5*tanh((274.15-Tabs)*0.3333))))
        rhonewice=rhonewsn
!--- compute density of "snowfall" from weighted contribution
!                 of snow, graupel and ice fractions

        **rhosnfall** = min(500.,max(58.8,(rhonewsn*snowrat +  &
!13mar18         rhosnfall = min(500.,max(76.9,(rhonewsn*snowrat +  &
                     rhonewgr*grauprat + rhonewice*icerat + rhonewgr*curat)))

! from now on rhonewsn is the density of falling frozen precipitation
        rhonewsn=rhosnfall
      end if

ericaligo-NOAA replied:
Also, my diag_tables include rhonewsn, not rhosnf
"gfs_phys",    "rhonewsn",   "rhonewsn",   "fv3_history2d",  "all",  .false.,  "none",  2

grantfirl replied:
Regarding the rhosnf question within module_sf_ruclsm, it is always only a diagnostic, but you're right that there are instances where it will be set to nonsense. 

1. Let's look at when exticeden == T.  `rhosnf` is intent(out) in the lsmruc subroutine, so it should always be set to something, but it is not. It is only initialized to  -1.e3 conditionally since it is in the `if(init .and. (lsm_cold_start) .and. iter == 1) then` block. `rhosnfall` is initialized to rhosnf (which is _uninitialized_ except for the first time step) prior to calling SFCTMP. Then, it is unmodified in SFCTMP and is used to set `rhosnf` upon return to lsmruc. So, this variable is an uninitialized diagnostic when exticeden == T. Perhaps this is OK since the externally-calculated variable (the one that is used) still has its value and the internal one is meaningless at this point.
2. Let's look at when exticeden == F. The same situation applies except that `rhosnfall` is conditionally given a value in SFCTMP. Otherwise, `rhosnf` returns nonsense. 

Critically, though, `rhosnf` potentially returning nonsense was the behavior **_before_** any of the modifications in this PR. It only ever had a value for the following conditions:

- `if(init .and. (lsm_cold_start) .and. iter == 1)` in lsmruc
- `IF((XLAND(I,J)-1.5) < 0.)` in lsmruc
- `IF(NEWSN.GT.0.)` in sfctmp

The initialization that happens to `rhosnfr` in the CCPP cap, lsm_ruc_run, is overridden by the intent(out) status of the variable in lsmruc.

I think that the concerns can be removed if `rhosnf` is declared as intent(inout) in lsmruc. That way, it should always at least have its value given at the first timestep. But, like I said, since it is only a diagnostic, it may not matter much.

grantfirl commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -4165,6 +4175,7 @@ subroutine control_initialize (Model, nlunit, fn_nml, me, master, &
     Model%ivegsrc          = ivegsrc
     Model%isot             = isot
     Model%use_ufo          = use_ufo
+    Model%exticeden        = exticeden
```
I put the following guard rail in my branch:`if (Model%exticeden .and. &
      (Model%imp_physics /= Model%imp_physics_gfdl .and. Model%imp_physics /= Model%imp_physics_thompson .and. &
       Model%imp_physics /= Model%imp_physics_nssl )) then`
      `!see GFS_MP_generic_post.F90; exticeden is only compatible with GFDL, Thompson, or NSSL MP `
      `print *,' Using exticeden = T is only valid when using GFDL, Thompson, or NSSL microphysics.'`
      `stop`
    `end if`

grantfirl replied:
I put this check in here because the calculation of rhonewsn in GFS_MP_generic_post is within an IF statement for these 3 MP schemes. If we don't have a check like this, it would be possible to specify exticeden = T in the namelist for a suite using a different microphysics and the LSMs would bypass their internal calculation of frozen precipitation density in favor of a bogus/initialized-only value of rhonewsn.

grantfirl replied:
You should also replace line 4200 with the following:
`if (Model%lsm==Model%lsm_noahmp .and. Model%exticeden .and. iopt_snf == 4) then`
 `      Model%iopt_snf         = 5`
   ` else`
     ` Model%iopt_snf         = iopt_snf`
    `end if`

grantfirl replied:
When NoahMP is used, the iopt_snf==5 option is used to cause the scheme to use the externally-calculated value of surface frozen precipitation ice density.

grantfirl replied:
You can view the changes that I'm talking about in GFS_typedefs.F90 here: https://github.com/ericaligo-NOAA/fv3atm/pull/1/files

ericaligo-NOAA replied:
I agree with all of your changes. See my earlier comments and concern about rhosnfall.

ericaligo-NOAA commented:
Thanks for the detailed analysis. I agree it's not doing anything at 
this point. I don't know the plans for the future if someone decided to 
use it after the call to SFCTMP where I saw:

RHOSNF(I,J) = RHOSNFALL (I,J).

It dies right there, but folks need to be careful about using the RHOSNF 
in the future. I'm fine with leaving it alone.


Eric

On 2/9/2023 1:58 PM, Grant Firl wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In ccpp/data/GFS_typedefs.F90 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#discussion_r1101901183>:
>
> > @@ -1729,6 +1730,14 @@ module GFS_typedefs
>       real (kind=kind_phys), pointer :: toticeb(:)     => null()   !< accumulated ice precipitation in bucket (kg/m2)
>       real (kind=kind_phys), pointer :: totsnwb(:)     => null()   !< accumulated snow precipitation in bucket (kg/m2)
>       real (kind=kind_phys), pointer :: totgrpb(:)     => null()   !< accumulated graupel precipitation in bucket (kg/m2)
> +    real (kind=kind_phys), pointer :: frzr   (:)     => null()   !< accumulated surface freezing rain (m)
> +    real (kind=kind_phys), pointer :: frzrb  (:)     => null()   !< accumulated surface freezing rain in bucket (m)
> +    real (kind=kind_phys), pointer :: frozr  (:)     => null()   !< accumulated surface graupel (m)
> +    real (kind=kind_phys), pointer :: frozrb (:)     => null()   !< accumulated surface graupel in bucket (m)
> +    real (kind=kind_phys), pointer :: tsnowp (:)     => null()   !< accumulated surface snowfall (m)
> +    real (kind=kind_phys), pointer :: tsnowpb(:)     => null()   !< accumulated surface snowfall in bucket (m)
> +    real (kind=kind_phys), pointer :: rhonewsn1(:)   => null()   !< precipitation ice density outside RUC LSM (kg/m3)
> +    real (kind=kind_phys), pointer :: rhosnf(:)      => null()   !< precipitation ice density inside RUC LSM (kg/m3)
>
> Regarding the rhosnf question within module_sf_ruclsm, it is always 
> only a diagnostic, but you're right that there are instances where it 
> will be set to nonsense.
>
>  1. Let's look at when exticeden == T. |rhosnf| is intent(out) in the
>     lsmruc subroutine, so it should always be set to something, but it
>     is not. It is only initialized to -1.e3 conditionally since it is
>     in the |if(init .and. (lsm_cold_start) .and. iter == 1) then|
>     block. |rhosnfall| is initialized to rhosnf (which is
>     /uninitialized/ except for the first time step) prior to calling
>     SFCTMP. Then, it is unmodified in SFCTMP and is used to set
>     |rhosnf| upon return to lsmruc. So, this variable is an
>     uninitialized diagnostic when exticeden == T. Perhaps this is OK
>     since the externally-calculated variable (the one that is used)
>     still has its value and the internal one is meaningless at this point.
>  2. Let's look at when exticeden == F. The same situation applies
>     except that |rhosnfall| is conditionally given a value in SFCTMP.
>     Otherwise, |rhosnf| returns nonsense.
>
> Critically, though, |rhosnf| potentially returning nonsense was the 
> behavior */before/* any of the modifications in this PR. It only ever 
> had a value for the following conditions:
>
>   * |if(init .and. (lsm_cold_start) .and. iter == 1)| in lsmruc
>   * |IF((XLAND(I,J)-1.5) < 0.)| in lsmruc
>   * |IF(NEWSN.GT.0.)| in sfctmp
>
> The initialization that happens to |rhosnfr| in the CCPP cap, 
> lsm_ruc_run, is overridden by the intent(out) status of the variable 
> in lsmruc.
>
> I think that the concerns can be removed if |rhosnf| is declared as 
> intent(inout) in lsmruc. That way, it should always at least have its 
> value given at the first timestep. But, like I said, since it is only 
> a diagnostic, it may not matter much.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/NOAA-EMC/fv3atm/pull/614#discussion_r1101901183>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ALQ75MKXHHHLQWEKH3B3GW3WWU45RANCNFSM6AAAAAATMC6YIY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

ericaligo-NOAA reviewed (commented):
Looks good.

ericaligo-NOAA committed:
include Grant's winter wx mods

ericaligo-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into winterwx

ChunxiZhang-NOAA commented:
@grantfirl Could you make a final review to PR at your convenience?

jkbk2004 commented:
@ericaligo-NOAA ccpp pr was merged. Please, update submodule pointer and revert the chnage in gitmodules.

ericaligo-NOAA committed:
Revert .gitmodules and update pointer to ccpp/physics

ChunxiZhang-NOAA commented:
@jkbk2004 Done.

jkbk2004 commented:
> @jkbk2004 Done.

Thanks!
